code/modules/cm_marines/specialist.dm

```diff
@@ -38,3 +38,151 @@
 	for (var/datum/action/item_action/specialist/SA in H.actions)
 		if (SA.ability_primacy == SPEC_PRIMARY_ACTION_2)
 			SA.handle_spec_macro()
+
+/// Get a specialist set datum typepath given a mob, returns null if they aren't a spec or haven't chosen a kit.
+/proc/get_specialist_set(mob/living/spec)
+	for(var/datum/specialist_set/path as anything in GLOB.specialist_set_datums)
+		if(HAS_TRAIT(spec, TRAIT_SPEC(path::trait_to_give)))
+			return path
+
+/proc/setup_specialist_sets()
+	var/list/set_list = list()
+	for(var/datum/specialist_set/path as anything in subtypesof(/datum/specialist_set))
+		var/datum/specialist_set/object = new path
+		set_list[path] = object
+		GLOB.specialist_set_name_dict[object.get_name()] = object
+	return set_list
+
+/datum/specialist_set
+	/// Human-readable name for the specialist set
+	VAR_PROTECTED/name = "" as text
+	/// What is the role title that should go on ID cards
+	VAR_PROTECTED/role_name = "" as text
+	/// How many more of this spec set can be picked from spec vendors
+	VAR_PRIVATE/available_vendor_num = 1 as num
+	/// How many more of this spec set can be picked from /obj/item/spec_kit
+	VAR_PRIVATE/available_kit_num = 2 as num
+	/// What skill tier to give the person redeeming the set
+	VAR_PROTECTED/skill_to_give = SKILL_SPEC_DEFAULT as num
+	/// What trait to give the person redeeming the set
+	VAR_PROTECTED/trait_to_give
+	/// What typepath to spawn for the redeemer if from a kit
+	VAR_PROTECTED/kit_typepath
+	/// List of typepaths that are incompatible with this set, meaning it'll subtract 1 from their vendor/kit num as well
+	VAR_PROTECTED/list/incompatible_sets = list()
+
+/datum/specialist_set/New()
+	. = ..()
+	incompatible_sets += type
+
+/datum/specialist_set/proc/redeem_set(mob/living/carbon/human/redeemer, kit = FALSE)
+	SHOULD_CALL_PARENT(TRUE)
+	if(!redeemer)
+		return FALSE
+
+	if(kit && (available_kit_num <= 0))
+		to_chat(redeemer, SPAN_WARNING("No more kits of this type may be chosen."))
+		return FALSE
+	else if(!kit && (available_vendor_num <= 0))
+		to_chat(redeemer, SPAN_WARNING("That set is already taken."))
+		return FALSE
+
+	if(skill_to_give != SKILL_SPEC_DEFAULT)
+		redeemer.skills?.set_skill(SKILL_SPEC_WEAPONS, skill_to_give)
+
+	if(kit)
+		redeemer.put_in_any_hand_if_possible(new kit_typepath, FALSE)
+		for(var/path in incompatible_sets)
+			GLOB.specialist_set_datums[path].available_kit_num--
+		ADD_TRAIT(redeemer, TRAIT_SPEC_KIT, TRAIT_SOURCE_INHERENT)
+	else
+		for(var/path in incompatible_sets)
+			GLOB.specialist_set_datums[path].available_vendor_num--
+		ADD_TRAIT(redeemer, TRAIT_SPEC_VENDOR, TRAIT_SOURCE_INHERENT)
+	ADD_TRAIT(redeemer, TRAIT_SPEC(trait_to_give), TRAIT_SOURCE_INHERENT)
+	var/obj/item/card/id/idcard = redeemer.wear_id
```

Comment:
```suggestion
	var/obj/item/card/id/idcard = redeemer.get_idcard()
```

Reply:
Otherwise I think it looks good? not doing anything incase its *somehow* intentional